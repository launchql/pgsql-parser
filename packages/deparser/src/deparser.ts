import { Node } from '@pgsql/types';
import { SqlFormatter } from './utils/sql-formatter';
import { ConsolidatedVisitor } from './visitors/consolidated-visitor';
import { DeparserContext } from './visitors/base';

export interface DeparserOptions {
  newline?: string;
  tab?: string;
}

export class Deparser {
  private formatter: SqlFormatter;
  private consolidatedVisitor: ConsolidatedVisitor;
  private tree: Node[];

  constructor(tree: Node | Node[], opts: DeparserOptions = {}) {
    this.formatter = new SqlFormatter(opts.newline, opts.tab);
    this.consolidatedVisitor = new ConsolidatedVisitor(this.formatter, this);
    
    this.tree = Array.isArray(tree) ? tree : [tree];
  }

  static deparse(query: Node | Node[], opts: DeparserOptions = {}): string {
    return new Deparser(query, opts).deparseQuery();
  }

  deparseQuery(): string {
    return this.tree
      .map(node => this.deparse(node))
      .join(this.formatter.newline() + this.formatter.newline());
  }

  deparse(node: Node, context: DeparserContext = {}): string | null {
    if (node == null) {
      return null;
    }

    if (typeof node === 'number' || node instanceof Number) {
      return node.toString();
    }

    try {
      return this.consolidatedVisitor.visit(node, context);
    } catch (error) {
      const nodeType = Object.keys(node)[0];
      throw new Error(`Error deparsing ${nodeType}: ${(error as Error).message}`);
    }
  }


}
