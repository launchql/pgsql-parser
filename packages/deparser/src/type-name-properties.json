{
  "TypeCast": ["typeName"],
  "RangeTableFuncCol": ["typeName"],
  "ColumnDef": ["typeName"],
  "XmlSerialize": ["typeName"],
  "JsonOutput": ["typeName"],
  "JsonTableColumn": ["typeName"],
  "CreateDomainStmt": ["typeName"],
  "CreateTransformStmt": ["type_name"],
  "CreateStmt": ["ofTypename"],
  "CreateOpClassItem": ["datatype", "storedtype"],
  "CreateFunctionStmt": ["returnType"],
  "FunctionParameter": ["argType"],
  "CreateCastStmt": ["sourcetype", "targettype"]
} 