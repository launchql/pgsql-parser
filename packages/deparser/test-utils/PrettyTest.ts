import { expectParseDeparse } from './index';

const generateCoded = require('../../../__fixtures__/generated/generated.json');

export class PrettyTest {
  private testCases: string[];
  private testPrefix: string;

  constructor(testCases: string[], testPrefix: string = 'test') {
    this.testCases = testCases;
    this.testPrefix = testPrefix;
  }

  /**
   * Generate individual tests for each test case with both pretty and non-pretty formatting
   */
  generateTests(): void {
    this.testCases.forEach((key, index) => {
      const sql = generateCoded[key];
      const testName = `${this.testPrefix}-${index + 1}`;

      it(`should format ${testName}: ${key} (pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: true });
        expect(result).toMatchSnapshot();
      });

      it(`should format ${testName}: ${key} (non-pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: false });
        expect(result).toMatchSnapshot();
      });
    });
  }

  /**
   * Generate tests with custom test name extraction
   */
  generateTestsWithCustomNaming(nameExtractor: (key: string, index: number) => string): void {
    this.testCases.forEach((key, index) => {
      const sql = generateCoded[key];
      const testName = nameExtractor(key, index);

      it(`should format ${testName}: ${key} (pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: true });
        expect(result).toMatchSnapshot();
      });

      it(`should format ${testName}: ${key} (non-pretty)`, async () => {
        const result = await expectParseDeparse(sql, { pretty: false });
        expect(result).toMatchSnapshot();
      });
    });
  }

  /**
   * Generate a single AST validation test for all test cases
   */
  generateASTValidationTest(): void {
    it('should validate AST equivalence for all test cases', async () => {
      const allSql = this.testCases.map((key) => generateCoded[key]);
      
      for (const sql of allSql) {
        await expectParseDeparse(sql, { pretty: false });
        await expectParseDeparse(sql, { pretty: true });
      }
    });
  }
} 