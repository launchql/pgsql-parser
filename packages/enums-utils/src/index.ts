import toInt from './enums2int.json';
import toStr from './enums2str.json';
import nodes from './nodes.json';

export { toInt, toStr, nodes };
export const getEnum = (enumType, key) => {
  if (typeof key === 'number') {
    return toStr[enumType][key];
  } else {
    return toInt[enumType][key];
  }
};

export * from './types';
