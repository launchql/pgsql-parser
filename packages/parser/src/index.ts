import { Deparser, deparse } from 'pgsql-deparser';
import {
  parseQuery,
  parseQuerySync,
  parsePlPgSQLSync as parseFunction
} from 'libpg-query';

function mapStmt({ stmt, stmt_len, stmt_location }) {
  return {
    RawStmt: {
      stmt,
      stmt_len,
      stmt_location: stmt_location || 0
    }
  };
}

export const parse = (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = parseQuerySync(sql);
  return result.stmts.map(mapStmt);
};

export const parseAsync = async (sql) => {
  if (!sql) throw new Error('no SQL provided to parser');
  const result = await parseQuery(sql);
  return result.stmts.map(mapStmt);
};

export { deparse, Deparser, parseFunction };
