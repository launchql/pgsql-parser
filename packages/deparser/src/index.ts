import { Deparser } from "./deparser";

const deparse = Deparser.deparse;
export { deparse, Deparser };
