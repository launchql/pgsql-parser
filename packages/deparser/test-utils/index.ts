import { parse } from '@pgsql/parser';
import { deparse } from '../src';
import { cleanTree, cleanLines } from '../src/utils';
import { readFileSync } from 'fs';
import * as path from 'path';
import { expect } from '@jest/globals';
import {diff} from 'jest-diff'

type ParseErrorType = 
  | 'PARSE_FAILED'
  | 'INVALID_STATEMENT'
  | 'REPARSE_FAILED'
  | 'AST_MISMATCH'
  | 'UNEXPECTED_ERROR'
  | 'INVALID_DEPARSED_SQL';

interface ParseError extends Error {
  type: ParseErrorType;
  relativePath: string;
  sql: string;
  deparsedSql?: string;
  originalAst?: any;
  reparsedAst?: any;
  parseError?: string;
}

function createParseError(
  type: ParseErrorType,
  relativePath: string,
  sql: string,
  deparsedSql?: string,
  originalAst?: any,
  reparsedAst?: any,
  parseError?: string
): ParseError {
  const error = new Error(getErrorMessage(type)) as ParseError;
  error.type = type;
  error.relativePath = relativePath;
  error.sql = sql;
  error.deparsedSql = deparsedSql;
  error.originalAst = originalAst;
  error.reparsedAst = reparsedAst;
  error.parseError = parseError;
  return error;
}

function getErrorMessage(type: ParseErrorType): string {
  switch (type) {
    case 'PARSE_FAILED':
      return 'Parse failed - no statements returned';
    case 'INVALID_STATEMENT':
      return 'Invalid statement structure';
    case 'REPARSE_FAILED':
      return 'Reparse failed - no statements returned';
    case 'AST_MISMATCH':
      return 'AST mismatch after parse/deparse cycle';
    case 'UNEXPECTED_ERROR':
      return 'Unexpected error during parse/deparse cycle';
    case 'INVALID_DEPARSED_SQL':
      return 'Invalid deparsed SQL';
  }
}

export class TestUtils {
  protected printErrorMessage(sql: string, position: number) {
    const lineNumber = sql.slice(0, position).match(/\n/g)?.length || 0;
    const lines = sql.split('\n');
    let colNumber = position - 1;
    for (let l = 0; l < lineNumber; l++) {
      colNumber -= lines[l].length + 1;
    }
    const errMessage = [`Error line ${lineNumber + 1}, column ${colNumber + 1}`];
    if (lineNumber > 0) {
      errMessage.push(lines[lineNumber - 1]);
    }
    errMessage.push(lines[lineNumber]);
    errMessage.push(' '.repeat(colNumber) + '^');
    if (lineNumber < lines.length - 1) {
      errMessage.push(lines[lineNumber + 1]);
    }
    console.error(errMessage.join('\n'));
  }

  tryParse(sql: string) {
    try {
      return parse(sql);
    } catch (err: any) {
      if (err.cursorPosition) {
        this.printErrorMessage(sql, err.cursorPosition);
      }
      throw err;
    }
  }

  expectAstMatch(relativePath: string, sql: string) {
    let tree: any;
    try {
      tree = this.tryParse(sql);
      if (tree.stmts) {
        tree.stmts.forEach((stmt: any) => {
          if (stmt.stmt) {
            const outSql = deparse(stmt.stmt);
            let reparsed;
            try {
              reparsed = parse(outSql);
            } catch (parseErr) {
              throw createParseError(
                'INVALID_DEPARSED_SQL',
                relativePath,
                sql,
                outSql,
                cleanTree([stmt]),
                undefined,
                parseErr instanceof Error ? parseErr.message : String(parseErr)
              );
            }
            const originalClean = cleanTree([stmt]);
            const reparsedClean = cleanTree(reparsed.stmts || []);
            
            if (!tree.stmts) {
              throw createParseError('PARSE_FAILED', relativePath, sql);
            }
            
            if (!stmt.stmt) {
              throw createParseError('INVALID_STATEMENT', relativePath, sql, undefined, cleanTree(tree));
            }
            
            if (!reparsed.stmts) {
              throw createParseError('REPARSE_FAILED', relativePath, sql, outSql, originalClean);
            }
            
            try {
              expect(reparsedClean).toEqual(originalClean);
            } catch (err) {
              throw createParseError('AST_MISMATCH', relativePath, sql, outSql, originalClean, reparsedClean);
            }
          }
        });
      }
    } catch (err) {
      const errorMessages: string[] = [];
      
      if (err instanceof Error && 'type' in err) {
        const parseError = err as ParseError;
        errorMessages.push(`\n❌ ${parseError.type}: ${parseError.relativePath}`);
        errorMessages.push(`❌ INPUT SQL: ${parseError.sql}`);
        
        if (parseError.deparsedSql) {
          errorMessages.push(`❌ DEPARSED SQL: ${parseError.deparsedSql}`);
        }
        
        if (parseError.type === 'INVALID_DEPARSED_SQL') {
          errorMessages.push(
            `\n❌ DEPARSER GENERATED INVALID SQL:`,
            `ORIGINAL AST:`,
            JSON.stringify(parseError.originalAst, null, 2),
            `\nPARSE ERROR: ${parseError.parseError}`
          );
        } else if (parseError.type === 'AST_MISMATCH') {
          errorMessages.push(
            `\n❌ AST COMPARISON:`,
            `EXPECTED AST:`,
            JSON.stringify(parseError.originalAst, null, 2),
            `\nACTUAL AST:`,
            JSON.stringify(parseError.reparsedAst, null, 2),
            `\nDIFF (what's missing from actual vs expected):`,
            diff(parseError.originalAst, parseError.reparsedAst)
          );
        } else if (parseError.originalAst) {
          errorMessages.push(`❌ AST: ${JSON.stringify(parseError.originalAst, null, 2)}`);
        }
      } else {
        // Handle unexpected errors
        errorMessages.push(
          `\n❌ UNEXPECTED ERROR: ${relativePath}`,
          `❌ INPUT SQL: ${sql}`,
          `❌ ERROR: ${err instanceof Error ? err.message : String(err)}`
        );
        if (tree) {
          errorMessages.push(`❌ PARSED AST: ${JSON.stringify(cleanTree(tree), null, 2)}`);
        }
      }
      
      console.log(errorMessages.join('\n'));
      throw err;
    }
  }
}

export class FixtureTestUtils extends TestUtils {
  private fixtures: Record<string, string>;

  constructor() {
    super();
    const GENERATED_JSON = path.join(__dirname, '../../../__fixtures__/generated/generated.json');
    this.fixtures = JSON.parse(readFileSync(GENERATED_JSON, 'utf-8'));
  }

  getTestEntries(filters: string[]) {
    if (filters.length === 0) {
      return Object.entries(this.fixtures);
    }
    
    return Object.entries(this.fixtures).filter(([relPath]) => 
      filters.includes(relPath)
    );
  }

  runFixtureTests(filters: string[]) {
    this.getTestEntries(filters).forEach(([relativePath, sql]) => {
      try {
        this.expectAstMatch(relativePath, sql);
      } catch (err) {
        throw err;
      }
    });
  }
}
