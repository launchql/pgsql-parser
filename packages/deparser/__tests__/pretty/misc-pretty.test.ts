import { deparseSync } from '../../src';
import { parse } from 'libpg-query';
import { expectParseDeparse } from '../../test-utils';

describe('Pretty Misc SQL formatting', () => {
  const misc1Sql = `WITH recent_orders AS (
  SELECT o.id, o.user_id, o.created_at
  FROM orders o
  WHERE o.created_at > NOW() - INTERVAL '30 days'
), high_value_orders AS (
  SELECT r.user_id, COUNT(*) AS order_count, SUM(oi.price * oi.quantity) AS total_spent
  FROM recent_orders r
  JOIN order_items oi ON r.id = oi.order_id
  GROUP BY r.user_id
)
SELECT u.id, u.name, h.total_spent
FROM users u
JOIN high_value_orders h ON u.id = h.user_id
WHERE h.total_spent > 1000
ORDER BY h.total_spent DESC`;

  const misc2Sql = `SELECT
  department,
  employee_id,
  COUNT(*) FILTER (WHERE status = 'active') OVER (PARTITION BY department) AS active_count,
  RANK() OVER (PARTITION BY department ORDER BY salary DESC) AS salary_rank
FROM employee_status
GROUP BY GROUPING SETS ((department), (department, employee_id))`;

  const misc3Sql = `SELECT u.id, u.name, j.key, j.value
FROM users u,
LATERAL jsonb_each_text(u.preferences) AS j(key, value)
WHERE j.key LIKE 'notif_%' AND j.value::boolean = true`;

  const misc4Sql = `SELECT p.id, p.title,
  CASE
    WHEN EXISTS (
      SELECT 1 FROM reviews r
      WHERE r.product_id = p.id AND r.rating >= 4
    ) THEN 'Popular'
    ELSE 'Unrated'
  END AS status
FROM products p
WHERE p.archived = false`;

  const misc5Sql = `WITH logs AS (
  SELECT id, payload::json->>'event' AS event, (payload::json->>'ts')::timestamp AS ts
  FROM event_log
  WHERE ts > NOW() - INTERVAL '7 days'
)
SELECT event, COUNT(*) AS freq
FROM (
  SELECT DISTINCT event, ts::date AS event_day
  FROM logs
) d
GROUP BY event
ORDER BY freq DESC`;

  const misc6Sql = `SELECT 
  o.id AS order_id,
  u.name AS user_name,
  p.name AS product_name,
  s.status,
  sh.shipped_at,
  r.refund_amount
FROM orders o
JOIN users u 
  ON o.user_id = u.id
JOIN order_items oi 
  ON oi.order_id = o.id
JOIN products p 
  ON (
    (p.id = oi.product_id AND p.available = true)
    OR 
    (p.sku = oi.product_sku AND (p.discontinued = false OR p.replacement_id IS NOT NULL))
  )
LEFT JOIN shipping sh 
  ON (
    sh.order_id = o.id 
    AND (
      (sh.carrier = 'UPS' AND sh.tracking_number IS NOT NULL)
      OR 
      (sh.carrier = 'FedEx' AND sh.shipped_at > o.created_at + INTERVAL '1 day')
    )
  )
LEFT JOIN statuses s 
  ON s.id = o.status_id 
  AND (
    s.name != 'cancelled'
    OR (s.name = 'cancelled' AND s.updated_at > NOW() - INTERVAL '7 days')
  )
LEFT JOIN refunds r 
  ON r.order_id = o.id 
  AND (
    (r.status = 'approved' AND r.processed_at IS NOT NULL)
    OR 
    (r.status = 'pending' AND r.requested_at < NOW() - INTERVAL '14 days')
  )
WHERE o.created_at > NOW() - INTERVAL '90 days'
  AND u.active = true
  AND (
    s.status = 'shipped' 
    OR (
      s.status = 'processing' 
      AND EXISTS (
        SELECT 1 FROM order_notes n WHERE n.order_id = o.id AND n.note ILIKE '%expedite%'
      )
    )
  )
ORDER BY o.created_at DESC`;

  it('should format misc-1: Complex CTE with joins and aggregation (pretty)', async () => {
    const result = await expectParseDeparse(misc1Sql, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-1: Complex CTE with joins and aggregation (non-pretty)', async () => {
    const result = await expectParseDeparse(misc1Sql, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-2: Window functions with FILTER and GROUPING SETS (pretty)', async () => {
    const result = await expectParseDeparse(misc2Sql, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-2: Window functions with FILTER and GROUPING SETS (non-pretty)', async () => {
    const result = await expectParseDeparse(misc2Sql, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-3: LATERAL joins with JSON functions (pretty)', async () => {
    const result = await expectParseDeparse(misc3Sql, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-3: LATERAL joins with JSON functions (non-pretty)', async () => {
    const result = await expectParseDeparse(misc3Sql, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-4: EXISTS with nested subqueries and CASE (pretty)', async () => {
    const result = await expectParseDeparse(misc4Sql, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-4: EXISTS with nested subqueries and CASE (non-pretty)', async () => {
    const result = await expectParseDeparse(misc4Sql, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-5: Nested CTEs with type casts and subqueries (pretty)', async () => {
    const result = await expectParseDeparse(misc5Sql, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-5: Nested CTEs with type casts and subqueries (non-pretty)', async () => {
    const result = await expectParseDeparse(misc5Sql, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-6: Complex multi-table joins with nested conditions (pretty)', async () => {
    const result = await expectParseDeparse(misc6Sql, { pretty: true });
    expect(result).toMatchSnapshot();
  });

  it('should format misc-6: Complex multi-table joins with nested conditions (non-pretty)', async () => {
    const result = await expectParseDeparse(misc6Sql, { pretty: false });
    expect(result).toMatchSnapshot();
  });

  it('should validate AST equivalence for all misc cases', async () => {
    const testCases = [misc1Sql, misc2Sql, misc3Sql, misc4Sql, misc5Sql, misc6Sql];
    
    for (const sql of testCases) {
      await expectParseDeparse(sql, { pretty: false });
      await expectParseDeparse(sql, { pretty: true });
    }
  });
});
