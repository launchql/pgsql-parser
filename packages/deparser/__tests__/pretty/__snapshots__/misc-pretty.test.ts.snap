// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pretty Misc SQL formatting should format misc-1: Complex CTE with joins and aggregation 1`] = `
"WITH 
  recent_orders AS (SELECT
    o.id,
    o.user_id,
    o.created_at
  FROM orders AS o
  WHERE
    o.created_at > (now() - '30 days'::interval)),
  high_value_orders AS (SELECT
    r.user_id,
    count(*) AS order_count,
    sum(oi.price * oi.quantity) AS total_spent
  FROM recent_orders AS r
  JOIN order_items AS oi ON r.id = oi.order_id
  GROUP BY
    r.user_id)
SELECT
  u.id,
  u.name,
  h.total_spent
FROM users AS u
JOIN high_value_orders AS h ON u.id = h.user_id
WHERE
  h.total_spent > 1000
ORDER BY
  h.total_spent DESC"
`;

exports[`Pretty Misc SQL formatting should format misc-2: Window functions with FILTER and GROUPING SETS 1`] = `
"SELECT
  department,
  employee_id,
  count(*) FILTER (WHERE status = 'active') OVER (PARTITION BY department) AS active_count,
  rank() OVER (
    PARTITION BY department
    ORDER BY salary DESC
  ) AS salary_rank
FROM employee_status
GROUP BY
  GROUPING SETS (department, (department, employee_id))"
`;

exports[`Pretty Misc SQL formatting should format misc-3: LATERAL joins with JSON functions 1`] = `
"SELECT
  u.id,
  u.name,
  j.key,
  j.value
FROM users AS u, LATERAL jsonb_each_text(u.preferences) AS j(key, value)
WHERE
  j.key LIKE 'notif_%'
  AND CAST(j.value AS boolean) = true"
`;

exports[`Pretty Misc SQL formatting should format misc-4: EXISTS with nested subqueries and CASE 1`] = `
"SELECT
  p.id,
  p.title,
CASE 
  WHEN EXISTS (SELECT
    1
  FROM reviews AS r
  WHERE
    r.product_id = p.id
    AND r.rating >= 4) THEN 'Popular' 
  ELSE 'Unrated' 
END AS status
FROM products AS p
WHERE
  p.archived = false"
`;

exports[`Pretty Misc SQL formatting should format misc-5: Nested CTEs with type casts and subqueries 1`] = `
"WITH 
logs AS (SELECT
  id,
  CAST(payload AS pg_catalog.json) ->> 'event' AS event,
  CAST(CAST(payload AS pg_catalog.json) ->> 'ts' AS pg_catalog.timestamp) AS ts
FROM event_log
WHERE
  ts > (now() - '7 days'::interval))
SELECT
  event,
  count(*) AS freq
FROM ( SELECT DISTINCT
  event,
  ts::date AS event_day
FROM logs ) AS d
GROUP BY
  event
ORDER BY
  freq DESC"
`;

exports[`Pretty Misc SQL formatting should format misc-6: Complex multi-table joins with nested conditions 1`] = `
"SELECT
  o.id AS order_id,
  u.name AS user_name,
  p.name AS product_name,
  s.status,
  sh.shipped_at,
  r.refund_amount
FROM orders AS o
JOIN users AS u ON o.user_id = u.id
JOIN order_items AS oi ON oi.order_id = o.id
JOIN products AS p ON
  (p.id = oi.product_id
    AND p.available = true)
    OR (p.sku = oi.product_sku
    AND (p.discontinued = false
    OR p.replacement_id IS NOT NULL))
LEFT JOIN shipping AS sh ON sh.order_id = o.id
  AND ((sh.carrier = 'UPS'
  AND sh.tracking_number IS NOT NULL)
  OR (sh.carrier = 'FedEx'
  AND sh.shipped_at > (o.created_at + '1 day'::interval)))
LEFT JOIN statuses AS s ON s.id = o.status_id
  AND (s.name <> 'cancelled'
  OR (s.name = 'cancelled'
  AND s.updated_at > (now() - '7 days'::interval)))
LEFT JOIN refunds AS r ON r.order_id = o.id
  AND ((r.status = 'approved'
  AND r.processed_at IS NOT NULL)
  OR (r.status = 'pending'
  AND r.requested_at < (now() - '14 days'::interval)))
WHERE
  o.created_at > (now() - '90 days'::interval)
  AND u.active = true
  AND (s.status = 'shipped'
  OR (s.status = 'processing'
  AND EXISTS (SELECT
  1
FROM order_notes AS n
WHERE
  (n.order_id = o.id
  AND n.note ILIKE '%expedite%'))))
ORDER BY
  o.created_at DESC"
`;
